File: sdk/python/src/openlit/guard/_integration.py
"""
Auto-guard integration layer.

``setup_auto_guards`` wraps LLM provider methods a *second* time (after
the normal instrumentors) so that guards run on every call without any
changes to existing ``instrumentation/`` code.

Call chain after setup:

    User call
      -> Guard wrapper  (preflight → may deny/redact)
        -> Instrumentor wrapper  (OTel telemetry)
          -> Original SDK method  (actual API call)
        <- Instrumentor wrapper
      <- Guard wrapper  (postflight → may redact/warn)
    <- Returns to user

**Streaming limitation**: postflight guards require a complete response
object with ``choices[].message.content`` (or equivalent). Streaming
responses yield incremental chunks that extractors cannot fully
reassemble, so postflight guards are silently skipped for streamed
completions. Preflight guards always run.
"""

from __future__ import annotations

import logging
from typing import Any, Callable, Dict, List, Optional, Tuple

from wrapt import wrap_function_wrapper

from openlit.guard._base import (
    Guard,
    GuardAction,
    GuardDeniedError,
    PipelineResult,
)
from openlit.guard._pipeline import Pipeline

logger = logging.getLogger(__name__)


# ---------------------------------------------------------------------------
# Provider-specific text extractors
# ---------------------------------------------------------------------------


def _extract_openai_input(kwargs: Dict[str, Any]) -> str:
    messages = kwargs.get("messages") or kwargs.get("input") or []
    if isinstance(messages, str):
        return messages
    parts: List[str] = []
    for m in messages:
        if isinstance(m, dict):
            content = m.get("content", "")
            if isinstance(content, str):
                parts.append(content)
        elif isinstance(m, str):
            parts.append(m)
    return " ".join(parts)


def _extract_openai_output(response: Any) -> str:
    try:
        choices = getattr(response, "choices", None)
        if choices:
            return " ".join(
                getattr(getattr(c, "message", None), "content", "") or ""
                for c in choices
            )
        output = getattr(response, "output", None)
        if output:
            parts: List[str] = []
            for item in output:
                msg = getattr(item, "content", None)
                if isinstance(msg, list):
                    for block in msg:
                        text = getattr(block, "text", None)
                        if text:
                            parts.append(text)
                elif isinstance(msg, str):
                    parts.append(msg)
            return " ".join(parts)
    except Exception:
        pass
    return ""


def _extract_anthropic_input(kwargs: Dict[str, Any]) -> str:
    messages = kwargs.get("messages", [])
    parts: List[str] = []
    for m in messages:
        if isinstance(m, dict):
            content = m.get("content", "")
            if isinstance(content, str):
                parts.append(content)
            elif isinstance(content, list):
                for block in content:
                    if isinstance(block, dict) and block.get("type") == "text":
                        parts.append(block.get("text", ""))
    return " ".join(parts)


def _extract_anthropic_output(response: Any) -> str:
    try:
        content = getattr(response, "content", [])
        parts: List[str] = []
        for block in content:
            text = getattr(block, "text", None)
            if text:
                parts.append(text)
        return " ".join(parts)
    except Exception:
        return ""


def _extract_generic_input(kwargs: Dict[str, Any]) -> str:
    for key in ("messages", "message", "prompt", "input", "text"):
        val = kwargs.get(key)
        if val is None:
            continue
        if isinstance(val, str):
            return val
        if isinstance(val, list):
            parts: List[str] = []
            for item in val:
                if isinstance(item, str):
                    parts.append(item)
                elif isinstance(item, dict):
                    parts.append(item.get("content", "") or "")
            return " ".join(parts)
    return ""


def _extract_generic_output(response: Any) -> str:
    try:
        choices = getattr(response, "choices", None)
        if choices:
            return " ".join(
                getattr(getattr(c, "message", None), "content", "") or ""
                for c in choices
            )
    except Exception:
        pass
    try:
        content = getattr(response, "content", None)
        if isinstance(content, str):
            return content
        if isinstance(content, list):
            parts: List[str] = []
            for block in content:
                text = getattr(block, "text", None)
                if text:
                    parts.append(text)
            if parts:
                return " ".join(parts)
    except Exception:
        pass
    try:
        text = getattr(response, "text", None)
        if text:
            return text
    except Exception:
        pass
    return ""


# ---------------------------------------------------------------------------
# Methods to guard:  (module, class.method, input_extractor, output_extractor)
# Only text-in / text-out methods; skip embeddings, images, audio, etc.
# ---------------------------------------------------------------------------

Extractor = Callable[..., str]

GUARDED_METHODS: List[Tuple[str, str, Extractor, Extractor]] = [
    # OpenAI
    (
        "openai.resources.chat.completions",
        "Completions.create",
        _extract_openai_input,
        _extract_openai_output,
    ),
    (
        "openai.resources.chat.completions",
        "AsyncCompletions.create",
        _extract_openai_input,
        _extract_openai_output,
    ),
    (
        "openai.resources.chat.completions",
        "Completions.parse",
        _extract_openai_input,
        _extract_openai_output,
    ),
    (
        "openai.resources.chat.completions",
        "AsyncCompletions.parse",
        _extract_openai_input,
        _extract_openai_output,
    ),
    (
        "openai.resources.responses.responses",
        "Responses.create",
        _extract_openai_input,
        _extract_openai_output,
    ),
    (
        "openai.resources.responses.responses",
        "AsyncResponses.create",
        _extract_openai_input,
        _extract_openai_output,
    ),
    # Anthropic
    (
        "anthropic.resources.messages",
        "Messages.create",
        _extract_anthropic_input,
        _extract_anthropic_output,
    ),
    (
        "anthropic.resources.messages",
        "AsyncMessages.create",
        _extract_anthropic_input,
        _extract_anthropic_output,
    ),
    # Groq
    (
        "groq.resources.chat.completions",
        "Completions.create",
        _extract_generic_input,
        _extract_generic_output,
    ),
    (
        "groq.resources.chat.completions",
        "AsyncCompletions.create",
        _extract_generic_input,
        _extract_generic_output,
    ),
    # Mistral
    (
        "mistralai.chat",
        "Chat.complete",
        _extract_generic_input,
        _extract_generic_output,
    ),
    (
        "mistralai.chat",
        "Chat.complete_async",
        _extract_generic_input,
        _extract_generic_output,
    ),
    # Cohere
    (
        "cohere.client_v2",
        "ClientV2.chat",
        _extract_generic_input,
        _extract_generic_output,
    ),
    (
        "cohere.client_v2",
        "AsyncClientV2.chat",
        _extract_generic_input,
        _extract_generic_output,
    ),
    # Together
    (
        "together.resources.chat.completions",
        "Completions.create",
        _extract_generic_input,
        _extract_generic_output,
    ),
    (
        "together.resources.chat.completions",
        "AsyncCompletions.create",
        _extract_generic_input,
        _extract_generic_output,
    ),
]


# ---------------------------------------------------------------------------
# Guard wrapper factories
# ---------------------------------------------------------------------------


def _apply_preflight(
    pipeline: Pipeline,
    kwargs: Dict[str, Any],
    extract_input: Extractor,
) -> Tuple[Dict[str, Any], Optional[PipelineResult]]:
    """Run preflight guards. Returns (possibly-modified kwargs, result_or_None)."""
    input_text = extract_input(kwargs)
    if not input_text:
        return kwargs, None

    result = pipeline.evaluate(input_text, phase="preflight")

    if result.action == GuardAction.DENY:
        raise GuardDeniedError(result)

    if result.action == GuardAction.REDACT and result.transformed_text is not None:
        for key in ("messages", "input", "prompt", "text"):
            if key in kwargs:
                original = kwargs[key]
                if isinstance(original, str):
                    kwargs = {**kwargs, key: result.transformed_text}
                elif isinstance(original, list) and original:
                    new_messages = list(original)
                    last = new_messages[-1]
                    if isinstance(last, dict) and "content" in last:
                        new_messages[-1] = {**last, "content": result.transformed_text}
                        kwargs = {**kwargs, key: new_messages}
                break

    return kwargs, result


def _apply_postflight(
    pipeline: Pipeline,
    response: Any,
    extract_output: Extractor,
) -> Any:
    """Run postflight guards on the response. Applies REDACT to response content."""
    output_text = extract_output(response)
    if not output_text:
        return response

    result = pipeline.evaluate(output_text, phase="postflight")

    if result.action == GuardAction.DENY:
        raise GuardDeniedError(result)

    if result.action == GuardAction.REDACT and result.transformed_text is not None:
        try:
            choices = getattr(response, "choices", None)
            if choices:
                for choice in choices:
                    msg = getattr(choice, "message", None)
                    if msg and hasattr(msg, "content"):
                        msg.content = result.transformed_text
            else:
                content = getattr(response, "content", None)
                if isinstance(content, list) and content:
                    for block in content:
                        if hasattr(block, "text"):
                            block.text = result.transformed_text
                            break
                elif hasattr(response, "text"):
                    response.text = result.transformed_text
        except Exception:
            pass

    return response


def _make_sync_guard_wrapper(
    pipeline: Pipeline,
    extract_input: Extractor,
    extract_output: Extractor,
):
    def _guard_wrapper(wrapped, instance, args, kwargs):
        kwargs, _ = _apply_preflight(pipeline, kwargs, extract_input)
        response = wrapped(*args, **kwargs)
        response = _apply_postflight(pipeline, response, extract_output)
        return response

    return _guard_wrapper


def _make_async_guard_wrapper(
    pipeline: Pipeline,
    extract_input: Extractor,
    extract_output: Extractor,
):
    async def _async_guard_wrapper(wrapped, instance, args, kwargs):
        kwargs, _ = _apply_preflight(pipeline, kwargs, extract_input)
        response = await wrapped(*args, **kwargs)
        response = _apply_postflight(pipeline, response, extract_output)
        return response

    return _async_guard_wrapper


# ---------------------------------------------------------------------------
# Public entry point
# ---------------------------------------------------------------------------


def setup_auto_guards(
    guards: List[Guard],
    fail_open: bool = True,
) -> None:
    """
    Wrap all known LLM provider methods with guard logic.

    Called from ``openlit.init()`` *after* all normal instrumentors have run.
    """
    pipeline = Pipeline(guards=guards, fail_open=fail_open)

    from openlit._config import OpenlitConfig

    OpenlitConfig.guard_pipeline = pipeline

    wrapped_count = 0
    for module_path, class_method, extract_in, extract_out in GUARDED_METHODS:
        is_async = "Async" in class_method
        try:
            if is_async:
                wrapper = _make_async_guard_wrapper(pipeline, extract_in, extract_out)
            else:
                wrapper = _make_sync_guard_wrapper(pipeline, extract_in, extract_out)

            wrap_function_wrapper(module_path, class_method, wrapper)
            wrapped_count += 1
        except Exception:
            logger.debug(
                "Could not wrap %s.%s with guard — module likely not installed",
                module_path,
                class_method,
            )

    logger.info(
        "Auto-guards: wrapped %d/%d provider methods",
        wrapped_count,
        len(GUARDED_METHODS),
    )

File: sdk/typescript/src/guard/integration.ts
/**
 * Auto-guard integration layer.
 *
 * setupAutoGuards wraps LLM provider methods so that guards run on every
 * call without any changes to existing instrumentation code.
 *
 * Call chain after setup:
 *
 *   User call
 *     -> Guard wrapper  (preflight -> may deny/redact)
 *       -> Instrumentor wrapper  (OTel telemetry)
 *         -> Original SDK method  (actual API call)
 *       <- Instrumentor wrapper
 *     <- Guard wrapper  (postflight -> may redact/warn)
 *   <- Returns to user
 *
 * **Streaming limitation**: postflight guards require a complete response
 * object with `choices[].message.content` (or equivalent). Streaming
 * responses yield incremental chunks that extractors cannot fully
 * reassemble, so postflight guards are silently skipped for streamed
 * completions. Preflight guards always run.
 *
 * Must stay in sync with: sdk/python/src/openlit/guard/_integration.py
 */

import { Guard, GuardAction, GuardDeniedError } from './base';
import { Pipeline } from './pipeline';

// ---------------------------------------------------------------------------
// Provider-specific text extractors
// ---------------------------------------------------------------------------

type Extractor = (arg: any) => string;

export function extractOpenAIInput(kwargs: Record<string, any>): string {
  const messages = kwargs.messages || kwargs.input || [];
  if (typeof messages === 'string') return messages;
  const parts: string[] = [];
  if (Array.isArray(messages)) {
    for (const m of messages) {
      if (typeof m === 'object' && m !== null) {
        const content = m.content ?? '';
        if (typeof content === 'string') parts.push(content);
      } else if (typeof m === 'string') {
        parts.push(m);
      }
    }
  }
  return parts.join(' ');
}

export function extractOpenAIOutput(response: any): string {
  try {
    const choices = response?.choices;
    if (choices && Array.isArray(choices)) {
      return choices
        .map((c: any) => c?.message?.content || '')
        .join(' ');
    }
    const output = response?.output;
    if (output && Array.isArray(output)) {
      const parts: string[] = [];
      for (const item of output) {
        const content = item?.content;
        if (Array.isArray(content)) {
          for (const block of content) {
            const text = block?.text;
            if (text) parts.push(text);
          }
        } else if (typeof content === 'string') {
          parts.push(content);
        }
      }
      return parts.join(' ');
    }
  } catch {
    // ignore
  }
  return '';
}

export function extractAnthropicInput(kwargs: Record<string, any>): string {
  const messages = kwargs.messages || [];
  const parts: string[] = [];
  if (Array.isArray(messages)) {
    for (const m of messages) {
      if (typeof m === 'object' && m !== null) {
        const content = m.content ?? '';
        if (typeof content === 'string') {
          parts.push(content);
        } else if (Array.isArray(content)) {
          for (const block of content) {
            if (typeof block === 'object' && block?.type === 'text') {
              parts.push(block.text || '');
            }
          }
        }
      }
    }
  }
  return parts.join(' ');
}

export function extractAnthropicOutput(response: any): string {
  try {
    const content = response?.content;
    if (Array.isArray(content)) {
      const parts: string[] = [];
      for (const block of content) {
        const text = block?.text;
        if (text) parts.push(text);
      }
      return parts.join(' ');
    }
  } catch {
    // ignore
  }
  return '';
}

export function extractGenericInput(kwargs: Record<string, any>): string {
  for (const key of ['messages', 'message', 'prompt', 'input', 'text']) {
    const val = kwargs[key];
    if (val == null) continue;
    if (typeof val === 'string') return val;
    if (Array.isArray(val)) {
      const parts: string[] = [];
      for (const item of val) {
        if (typeof item === 'string') {
          parts.push(item);
        } else if (typeof item === 'object' && item !== null) {
          parts.push(item.content || '');
        }
      }
      return parts.join(' ');
    }
  }
  return '';
}

export function extractGenericOutput(response: any): string {
  try {
    const choices = response?.choices;
    if (choices && Array.isArray(choices)) {
      return choices
        .map((c: any) => c?.message?.content || '')
        .join(' ');
    }
  } catch {
    // ignore
  }
  try {
    const content = response?.content;
    if (typeof content === 'string') return content;
    if (Array.isArray(content)) {
      const parts: string[] = [];
      for (const block of content) {
        const text = block?.text;
        if (text) parts.push(text);
      }
      if (parts.length) return parts.join(' ');
    }
  } catch {
    // ignore
  }
  try {
    if (response?.text) return response.text;
  } catch {
    // ignore
  }
  return '';
}

// ---------------------------------------------------------------------------
// Methods to guard: (npmPackage, protoPath, methodName, inputExtractor, outputExtractor)
// ---------------------------------------------------------------------------

interface GuardedMethod {
  npmPackage: string;
  protoPath: string[];
  methodName: string;
  extractInput: Extractor;
  extractOutput: Extractor;
}

const GUARDED_METHODS: GuardedMethod[] = [
  // OpenAI
  { npmPackage: 'openai', protoPath: ['OpenAI', 'Chat', 'Completions', 'prototype'], methodName: 'create', extractInput: extractOpenAIInput, extractOutput: extractOpenAIOutput },
  { npmPackage: 'openai', protoPath: ['OpenAI', 'Responses', 'prototype'], methodName: 'create', extractInput: extractOpenAIInput, extractOutput: extractOpenAIOutput },
  // Anthropic
  { npmPackage: '@anthropic-ai/sdk', protoPath: ['Anthropic', 'Messages', 'prototype'], methodName: 'create', extractInput: extractAnthropicInput, extractOutput: extractAnthropicOutput },
  // Groq
  { npmPackage: 'groq-sdk', protoPath: ['Groq', 'Chat', 'Completions', 'prototype'], methodName: 'create', extractInput: extractGenericInput, extractOutput: extractGenericOutput },
  // Mistral
  { npmPackage: '@mistralai/mistralai', protoPath: ['Mistral', 'chat', 'prototype'], methodName: 'complete', extractInput: extractGenericInput, extractOutput: extractGenericOutput },
  // Cohere
  { npmPackage: 'cohere-ai', protoPath: ['CohereClientV2', 'prototype'], methodName: 'chat', extractInput: extractGenericInput, extractOutput: extractGenericOutput },
  // Together
  { npmPackage: 'together-ai', protoPath: ['Together', 'Chat', 'Completions', 'prototype'], methodName: 'create', extractInput: extractGenericInput, extractOutput: extractGenericOutput },
  // Bedrock (AWS)
  { npmPackage: '@aws-sdk/client-bedrock-runtime', protoPath: ['BedrockRuntimeClient', 'prototype'], methodName: 'send', extractInput: extractGenericInput, extractOutput: extractGenericOutput },
  // Google Generative AI
  { npmPackage: '@google/generative-ai', protoPath: ['GenerativeModel', 'prototype'], methodName: 'generateContent', extractInput: extractGenericInput, extractOutput: extractGenericOutput },
];

// ---------------------------------------------------------------------------
// Preflight / Postflight
// ---------------------------------------------------------------------------

function applyPreflight(
  pipeline: Pipeline,
  kwargs: Record<string, any>,
  extractInput: Extractor,
): Record<string, any> {
  const inputText = extractInput(kwargs);
  if (!inputText) return kwargs;

  const result = pipeline.evaluate(inputText, 'preflight');

  if (result.action === GuardAction.DENY) {
    throw new GuardDeniedError(result);
  }

  if (result.action === GuardAction.REDACT && result.transformedText !== null) {
    for (const key of ['messages', 'input', 'prompt', 'text']) {
      if (key in kwargs) {
        const original = kwargs[key];
        if (typeof original === 'string') {
          return { ...kwargs, [key]: result.transformedText };
        }
        if (Array.isArray(original) && original.length > 0) {
          const newMessages = [...original];
          const last = newMessages[newMessages.length - 1];
          if (typeof last === 'object' && last !== null && 'content' in last) {
            newMessages[newMessages.length - 1] = { ...last, content: result.transformedText };
            return { ...kwargs, [key]: newMessages };
          }
        }
        break;
      }
    }
  }

  return kwargs;
}

function applyPostflight(
  pipeline: Pipeline,
  response: any,
  extractOutput: Extractor,
): any {
  const outputText = extractOutput(response);
  if (!outputText) return response;

  const result = pipeline.evaluate(outputText, 'postflight');

  if (result.action === GuardAction.DENY) {
    throw new GuardDeniedError(result);
  }

  if (result.action === GuardAction.REDACT && result.transformedText !== null) {
    try {
      const choices = response?.choices;
      if (choices && Array.isArray(choices)) {
        for (const choice of choices) {
          if (choice?.message && 'content' in choice.message) {
            choice.message.content = result.transformedText;
          }
        }
      } else {
        const content = response?.content;
        if (Array.isArray(content) && content.length > 0) {
          for (const block of content) {
            if ('text' in block) {
              block.text = result.transformedText;
              break;
            }
          }
        } else if (response && 'text' in response) {
          response.text = result.transformedText;
        }
      }
    } catch {
      // best-effort redaction
    }
  }

  return response;
}

// ---------------------------------------------------------------------------
// Prototype resolver
// ---------------------------------------------------------------------------

function resolveProto(moduleExports: any, protoPath: string[]): any {
  let obj = moduleExports;
  for (const key of protoPath) {
    if (obj == null) return null;
    obj = obj[key];
  }
  return obj;
}

// ---------------------------------------------------------------------------
// Public entry point
// ---------------------------------------------------------------------------

export function setupAutoGuards(
  guards: Guard[],
  failOpen = true,
): void {
  const pipeline = new Pipeline({ guards, failOpen });

  try {
    const OpenlitConfig = require('../config').default;
    OpenlitConfig.guardPipeline = pipeline;
  } catch {
    // config not available
  }

  let wrappedCount = 0;
  for (const { npmPackage, protoPath, methodName, extractInput, extractOutput } of GUARDED_METHODS) {
    try {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      const mod = require(npmPackage);
      const proto = resolveProto(mod, protoPath);
      if (!proto || typeof proto[methodName] !== 'function') continue;

      const original = proto[methodName];
      proto[methodName] = async function (this: any, ...args: any[]) {
        let params = args[0] || {};

        params = applyPreflight(pipeline, params, extractInput);
        args[0] = params;

        const response = await original.apply(this, args);

        return applyPostflight(pipeline, response, extractOutput);
      };
      wrappedCount++;
    } catch {
      // module not installed, skip silently
    }
  }

  if (wrappedCount > 0) {
    console.log(`[openlit] Auto-guards: wrapped ${wrappedCount}/${GUARDED_METHODS.length} provider methods`);
  }
}
